### ISSUE TITLE
Body transform doesn't seem to handle non-ascii characters

### ISSUE THREAD

**Issue by lonelycode** [TykTechnologies/tyk#642]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
I can use the transformation functions (body transform) to transform XML and JSON just fine, but it fails when using non-latin characters

**What is the expected behavior?**
THat it should work with all kinds of characters

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem**

See https://community.tyk.io/t/xml-to-json-transform-and-oa-characters/1524/3

**Which versions of Tyk affected by this issue? Did this work in previous versions of Tyk?**

All.

**Comment by mvdan** [comment]
I just tried to reproduce this using only the body transform middleware. Even when copying the user's stuff exactly as they posted it, it still works for me:

```
=== RUN   TestTransformNonAscii
{
    "Jyväskylä": "127.0.0.1",
    "Hyvinkää": "127.0.0.2"

}
```

So either this bug was fixed somehow (we don't know what version the user was running), or the bug is somewhere other than the middleware transform. Or perhaps this was somehow fixed in a Go version.

We haven't touched the `github.com/clbanning/mxj` recently, but I'm not sure if it was always vendored.

**Comment by lhbotha** [comment]
Hi,
I have the following versions installed:
tyk-dashboard 1.3.2 amd64
tyk-gateway 2.3.3 amd64

Br,
Louis

**Comment by mvdan** [comment]
@lhbotha are you certain it's the body transform? I assume you are because of the screenshots you pasted of the endpoint designer, and because you said the bug disappears if you remove the body transform.

Did you try the latest Tyk versions, namely 2.3.5 and 1.3.5 of the gateway and dashboard?

Can you give any more information on the API definition you're using? Are you using any other form of middleware or plugin with it?

Thank you!

**Comment by mvdan** [comment]
Friendly ping @lhbotha - do you have any more information you could share with us?

@ilijabojanovic will try to reproduce this for our upcoming 2.4 major release, but so far we haven't been able to reproduce it.

**Comment by ilijabojanovic** [comment]
@mvdan i was able to reproduce this issue over Dashboard. This is UI issue for @lghiur, since issue is input form on dashboard designer page.


Reproduction video:
https://screencast.com/t/OneODYnLWI

**Comment by mvdan** [comment]
Are you positive it was a UI issue? It could be a backend issue that I wasn't able to reproduce.

**Comment by ilijabojanovic** [comment]
@mvdan yes, its related to encoding/decoding strings in Dashboard UI 
https://github.com/angular-ui/ui-router/issues/421

**Comment by lghiur** [comment]
@ilijabojanovic I don't think that is the problem. For decoding the output I use the js function "atob" which should work ok, but the response I get from the server (from the test request http://localhost:3000/api/designer/test) looks like this:

```
{
	"Type": "xml",
	"Template": "ewp7e3JhbmdlICR4LCAkcyA6PSAuc2VydmVycy5zZXJ2ZXJ9fQoie3skcy5zZXJ2ZXJOYW1lfX0iOiAie3skcy5zZXJ2ZXJJUH19Igp7e2lmIG5vdCAkeH19LHt7ZW5kfX0Ke3tlbmR9fQp9",
	"Input": "PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0idXRmLTgiPz4KPHNlcnZlcnMgdmVyc2lvbj0iMSI+CiAgICA8c2VydmVyPgogICAgICAgIDxzZXJ2ZXJOYW1lPiBKeXbkc2t5bOQ8L3NlcnZlck5hbWU+CiAgICAgICAgPHNlcnZlcklQPjEyNy4wLjAuMTwvc2VydmVySVA+CiAgICA8L3NlcnZlcj4KICAgIDxzZXJ2ZXI+CiAgICAgICAgPHNlcnZlck5hbWU+dGVzdDI8L3NlcnZlck5hbWU+CiAgICAgICAgPHNlcnZlcklQPjEyNy4wLjAuMjwvc2VydmVySVA+CiAgICA8L3NlcnZlcj4KPC9zZXJ2ZXJzPg==",
	"Output": "ewoKfQ=="
}
```

Now if you try in the console this like of code:

`var a = window.atob('ewoKfQ=='); a;`

you will see that the string is the one in the output textarea "{ }".

Now I don't to throw this to the backend side so I will discuss with @mvdan about this to see how we can solve it.

**Comment by mvdan** [comment]
Thanks for confirming - one part of the codebase less to suspect. Will look at the backend again.

**Comment by lonelycode** [comment]
The atob function doesn't do UTF-8 well IIRC. We had some trouble a while back encoding blueprint docs for the portal that were in simplified chinese because we couldn't trust the atob func to handle it properly. 

https://stackoverflow.com/questions/30106476/using-javascripts-atob-to-decode-base64-doesnt-properly-decode-utf-8-strings

**Comment by mvdan** [comment]
Yeah, what I've come to see is that what the FE sends to the BE is not utf-8:

```
 $ base64 -d <<<PG5hbWVzPgogICAgICAgPG5hbWU+Snl25HNreWzkPC9uYW1lPgogICAgICAgPG5hbWU+SHl2aW5r5OQ8L25hbWU+CjwvbmFtZXM+ | file -
/dev/stdin: ISO-8859 text
```

Over to @lghiur again, for now :)

**Comment by mvdan** [comment]
So the good news is that this is a frontend issue and only affects the endpoint designer. The gateway should be able to handle non-ascii inputs in the body transform just fine - I checked with a unit test.

So for those wondering about a fix, for 2.3.X you can simply test your body transform template on the endpoint designer while skipping non-ascii characters in the input - they'll work when you actually configure the template with the gateway.

We'll include a fix in the endpoint designer for the upcoming major release.

**Comment by lghiur** [comment]
This is being fixed for v1.4 of the dashboard. The PR is here https://github.com/TykTechnologies/tyk-analytics-ui/pull/129

**Comment by ilijabojanovic** [comment]
Fixed on 1.4 release
